### PR TITLE
Remove support for delete markers and improve config consistency

### PR DESCRIPTION
**Pull request by jaydoane**
## Overview

Remove support for delete markers, and improve consistency of config values following deletion.

Calling the persistent variation of `config:delete("section", "key")` writes `key =` to the most significant `.ini` file, internally referred to as a "delete marker" which, [according to the docs](https://docs.couchdb.org/en/stable/config/intro.html#setting-parameters-via-the-configuration-file), should cause all previous values of `key` in the `.ini` file chain to be ignored, and sets the value to the atom `undefined`.

Currently, `config:reload` [ignores delete markers](https://github.com/apache/couchdb/issues/4892). Also, since `config:delete` doesn't reload values from `.ini` files at all, using the persistent variation may result in temporarily returning an inconsistent value until the `config` server is restarted.

This change removes support for delete markers. A persistent call to `config:delete` no longer writes a delete marker, but instead deletes the entire line. When parsing `.ini` files, any existing  delete markers are ignored.

Additionally, this change improves consistency after `config:delete` by loading any values from `.ini` files into the `ets` table subsequent to the deletion.

Additionally, this PR:
- Simplifies logging of sensitive values by introducing the `maybe_conceal/2` function.
- Cleans up the config application:
  - Removes an unused record field, export, include, and some defines.
  - Removes a deletion "reason" that was probably a copy/pasta error.
  - Fixes `should_create_non_persistent_option` and `should_create_persistent_option` tests so that they will run.
- Improves test coverage:

Before:
```
  All 59 tests passed.
Code Coverage:
config              :  84%
config_app          :  87%
config_listener     :  66%
config_listener_mon :  65%
config_notifier     :  93%
config_sup          : 100%
config_util         :   4%
config_writer       : 100%

Total               : 80%
```
After:
```
  All 63 tests passed.
Code Coverage:
config              :  87%
config_app          :  87%
config_listener     :  66%
config_listener_mon :  65%
config_notifier     :  93%
config_sup          : 100%
config_util         :  52%
config_writer       :  95%

Total               : 85%
```

## Testing recommendations

```
make check
```

## Checklist

- [x] Code is written and works correctly
- [x] Changes are covered by tests
- [ ] Any new configurable parameters are documented in `rel/overlay/etc/default.ini`
- [ ] Documentation changes were made in the `src/docs` folder
- [ ] Documentation changes were backported (separated PR) to affected branches

### CODE REVIEW

File: src/config/src/config.erl [hunk 1]
```diff
@@ -326,6 +326,13 @@ handle_call({delete, Sec, Key, Persist, Reason}, _From, Config) ->
         "~p: [~s] ~s deleted for reason ~p",
         [?MODULE, Sec, Key, Reason]
     ),
+    DiskKVs = parse_ini_files(Config#config.ini_files),
+    case dict:find({Sec, Key}, DiskKVs) of
+        {ok, DiskVal} ->
+            true = ets:insert(?MODULE, {{Sec, Key}, DiskVal});
+        error ->
+            ok
+    end,
```
Comment (nickva) [inline src/config/src/config.erl:335]: To emphasize that we're aiming to get the value that we'd get effectively after a node restart or `reload()`, it might make sense to do the parsing and reading after the write. That's just so we go through the same sequence of operations: 1) parse from ini and 2) update ETS, that we'd perform when we re-start the node or when we call `reload()`.

Comment (jaydoane) [reply src/config/src/config.erl]: Good idea, makes sense. Fixed in https://github.com/apache/couchdb/pull/4813/commits/bb0eb8fd3064ef8eaa7fc6a633e9d825e046427e

File: src/config/test/config_tests.erl [hunk 2]
```diff
@@ -441,8 +437,8 @@ should_write_changes(_, _) ->
         ?assertEqual("5986", config:get("httpd", "port")),
         ?assertEqual(ok, config:set("httpd", "port", "8080")),
         ?assertEqual("8080", config:get("httpd", "port")),
-        ?assertEqual(ok, config:delete("httpd", "bind_address", "8080")),
-        ?assertEqual(undefined, config:get("httpd", "bind_address"))
+        ?assertEqual(ok, config:delete("httpd", "bind_address")),
+        ?assertEqual("127.0.0.1", config:get("httpd", "bind_address"))
```
Comment (pgj) [inline src/config/test/config_tests.erl:441]: Why does not this remain `undefined`?

File: src/config/src/config.erl [hunk 3]
```diff
@@ -252,7 +251,7 @@ init(IniFiles) ->
     ets:new(?MODULE, [named_table, set, protected, {read_concurrency, true}]),
     lists:map(
         fun(IniFile) ->
-            {ok, ParsedIniValues} = parse_ini_file(IniFile),
+            {ok, ParsedIniValues} = parse_ini_file(IniFile, fun delete_keys/1),
```
Comment (nickva) [inline src/config/src/config.erl:254]: I wonder if with the new behavior  the initial load logic can be the same as the reload. Could we always call parse_init_files with the omit logic being the default. 

I tried to go through a few possible cases in a table:

|                      |    |    |    |    |  
| ------------ | - | - | -  | - | 
| default.ini     | X | D | D | X | 
| local.ini         | X | X | L | L | 
| active value  | C | D | L | L | 

(C - code default, X - delete marker, D, L - default and local values)

What do you think?

File: src/config/test/config_tests.erl (hunk 2 above)
Comment (jaydoane) [reply src/config/test/config_tests.erl:441]: It was `undefined` before because the `config:delete("httpd", "bind_address")` on the previous line removed it from the `ets` table, but did not reload the default. With this change, after deletion the default gets reloaded from disk and inserted into the ets table. In short, It is the crux of this PR.

File: src/config/src/config.erl (hunk 3 above)
Comment (jaydoane) [reply src/config/src/config.erl:254]: The only one I'm not entirely sure about is when there's a value in `default.ini`, but `local.ini` has a delete marker.  Shouldn't row 3 column 2 be the same as row 3 column 1, based on this paragraph in  [the docs](https://docs.couchdb.org/en/stable/config/intro.html#setting-parameters-via-the-configuration-file)?

> In case when you’d like to remove some parameter from the default.ini without modifying that file, you may override in local.ini, but without any value
```
[compactions]
_default =
```

But maybe I'm misunderstanding your table and/or the docs.

FWIW, it seems like eunit and elixir tests continue to pass if we replace `delete_keys/1` with `omit_delete/1` in the `init` function.

Comment (nickva) [reply src/config/src/config.erl:254]: Hmm, good point. I think the docs would apply more if we only had two config files: `default.ini` and `local.ini`, but since we support a hierarchy of files: `default.ini, default.d/*.ini, local.ini, local.d/*.ini` it's a bit ambiguous. What if the value in default.ini  is 'D', in local.ini as `L` and in local.ini/1.ini is a deleted entry? In the current main code we'd have a discrepancy:

 * During startup we'd end up with `C`, the code default (`undefined`)
 * After reload we'd end with `L`, the one higher in the hierarchy

It's seems complicated to reason about having startup and reload work differently. That discrepancy seems like a bug. I think it would be nice to at least get it to state where, absent any changes to the .ini files, a start-up followed by any number of `reload` calls shouldn't change the active config value.

In general, trying to think what is more useful from a user's point of view:
   * They have a `default.ini` and `local.ini`. They set some value in `local.ini`, say to increase `max_dbs_open = 1000` when their `default.ini` may have it at `500`. Then they decide they'd like to revert back to default.ini's value so they use the `delete` operation.
   * Always revert to code defaults from any point in the config hierarchy. Maybe in the case their default.ini is read-only or broken and code defaults are better. In that case, perhaps they could create another `local.d/*.ini` and write the good value there. (But if we do it this way, reload as is would behave inconsistently so we have to make it behave similarly to init at least - always go back to code default).

To me it seems the current reload behavior is more intuitive and the startup behavior seems surprising.

File: src/config/test/config_tests.erl (hunk 2 above)
Comment (pgj) [reply src/config/test/config_tests.erl:441]: Okay, that is what I thought.  Thanks for the confirmation.

File: src/config/src/config.erl (hunk 3 above)
Comment (jaydoane) [reply src/config/src/config.erl]: > I think the docs would apply more if we only had two config files: `default.ini` and `local.ini`, but since we support a hierarchy of files: `default.ini, default.d/*.ini, local.ini, local.d/*.ini` it's a bit ambiguous.

Yeah, I think the docs need to be updated to reflect the possibility of more `.ini` files. The change could be as simple as just:

"In case when you’d like to remove some parameter from an `.ini` file without modifying that file, you may override it in a subsequent `.ini` file, but without supplying any value following the equal sign:"


> What if the value in `default.ini` is 'D', in `local.ini` as `L` and in `local.ini/1.ini` is a deleted entry? In the current main code we'd have a discrepancy:
> 
>     * During startup we'd end up with `C`, the code default (`undefined`)
> 
>     * After reload we'd end with `L`, the one higher in the hierarchy

I believe this problem exists even with just a `default.ini` and `local.ini` file. I performed a little experiment in a dev environment on main that also showed a discrepancy between how `init` and `reload` treat delete markers. I started by prepending a new section and values to these files:
```diff
diff --git a/rel/overlay/etc/default.ini b/rel/overlay/etc/default.ini
index 68ae900a3..c1461662e 100644
--- a/rel/overlay/etc/default.ini
+++ b/rel/overlay/etc/default.ini
@@ -1,3 +1,6 @@
+[section]
+file = default.ini
+
 ; Upgrading CouchDB will overwrite this file.
 [vendor]
 name = {{package_author_name}}
diff --git a/rel/overlay/etc/local.ini b/rel/overlay/etc/local.ini
index ba76ebfdf..bd8c88565 100644
--- a/rel/overlay/etc/local.ini
+++ b/rel/overlay/etc/local.ini
@@ -1,3 +1,6 @@
+[section]
+file = local.ini
+
 ; CouchDB Configuration Settings

 ; Custom settings should be made in this file. They will override settings
```
then in a remsh after starting CouchDB:
```
(node1@127.0.0.1)2> config:get("section", "file").
"local.ini"
(node1@127.0.0.1)3> config:delete("section", "file").
ok
```
in another terminal, confirming `local.ini` was changed as expected:
```
❯ head -2 dev/lib/node1/etc/local.ini
[section]
file =
```
Now, this `undefined` below is actually "correct":
```
(node1@127.0.0.1)4> config:get("section", "file").
undefined
```
but for the wrong reason, whereas after a `reload`, it gives the wrong answer:
```
(node1@127.0.0.1)5> config:reload().
ok
(node1@127.0.0.1)6> config:get("section", "file").
"default.ini"
```
since the `file =` line in `local.ini` should have overridden the default, and made it `undefined`.

Lastly, I copied the modified `dev/lib/node1/etc/local.ini` to `rel/overlay/etc/local.ini` to prevent it being overwritten, and restarted CouchDB. Evidently, `config:init/1` correctly honors the delete marker:
```
(node1@127.0.0.1)1> config:get("section", "file").
undefined
```

> It's seems complicated to reason about having startup and reload work differently. The discrepancy itself seems like a bug.

I completely agree that it's hard to reason about, and clearly seems buggy based on the above.

> I think it would be nice to at least fix it to where, absent any changes to the .ini files, a start-up followed by any number of `reload` calls shouldn't change the active config value.

Agree 100%

> In general, trying to think what is more useful from a user's point of view:
> 
> * They have a `default.ini` and `local.ini`. They set some value in `local.ini`, say to increase `max_dbs_open = 1000` when their `default.ini` may have it at `500`. Then they decide they'd like to revert back to default.ini's value so they use the `delete` operation.

I think this behavior is reasonable, but I believe it would be a breaking change in how we interpret delete markers.

> * Always revert to code defaults from any point in the config hierarchy.

I'm not clear on what you mean here ^. Would this be the effect of ignoring delete markers?

> Maybe in the case their default.ini is read-only or broken and code defaults are better. In that case, perhaps they could create another `local.d/*.ini` and write the good value there.

I think the idea is that `default.ini` should always be treated as read-only because we reserve the right to change it with upgrades. So we heavily discourage admins from ever modifying `default.ini`, but instead make all overrides to `local.ini`. And a delete marker in any subsequent `.ini` file is a signifier that any previous value for that key should be ignored.

> (But if we do it this way, reload as is would behave inconsistently so we have to make it behave similarly to init at least - always go back to code default).

Yes, totally agree. I have attempted to do fix the discrepancy between `init` and `reload` in the latest update to this PR.

> To me it seems the current reload behavior is more intuitive and the startup behavior seems surprising.

Clearly the way `init` was making `ets:delete` calls when it hit delete markers was hard to reason about, and especially made the code hard to re-use.

The main recent changes are:
1. Implement `ini_map/1` leveraging the built in `maps:merge` to simplify parsing and merging multiple `.ini` files.
1. Make delete markers first class concepts, and expose their existence to any caller of `ini_map/1` so it can handle them accordingly.
1. Improve consistency of delete marker handling for `reload`, `init`, and `delete` using `ini_map/1`.

It could use some unit tests, but manual testing seems to now be giving consistent results.

File: src/config/test/config_tests.erl [hunk 4]
```diff
@@ -115,6 +111,41 @@ handle_config_terminate(Self, Reason, {Pid, State}) ->
...
+config_delete_reload_restart_test_() ->
+    {
+        "Test consistency after set, delete, reload, and restart",
+        foreachx,
+        fun setup_ini/1,
+        fun teardown_ini/2,
+        [
+            {
+                ["default.ini", "local.ini"],
+                fun non_persistent_set_delete_reload_restart/2
```
Comment (jaydoane) [inline src/config/test/config_tests.erl:140]: I tried adding these lines to `couch_eunit.hrl`:
```erlang
-ifndef(TDEF_FEX).
-define(TDEF_FEX(Name), fun(R, X) -> {atom_to_list(Name), ?_test(Name(R, X))} end).
-define(TDEF_FEX(Name, Timeout), fun(R, X) -> {atom_to_list(Name), {timeout, Timeout, ?_test(Name(R, X))}} end).
-endif.
```
and rewrote these new tests to use it.

However, as I worked on converting the _rest_ of this file to use `TDEF_FEX` in the `foreachx` tests, I realized it might take a bit more effort with tests like e.g. `should_notify_on_config_reload` which don't seem to lend themselves to a simple conversion. I will continue to ponder the best approach, and would welcome advice.

Comment (nickva) [reply src/config/test/config_tests.erl:140]: For one or two extra parameters, the complexity of a foreachx might not be worth it? For the two options just having two test setups: `setup_local_ini` and `setup_default_ini` and they can call the same setup `setup_ini/1` utility function could be simpler and could use the standard `?TDEF_FE` macro?

Comment (jaydoane) [reply src/config/test/config_tests.erl:140]: Those tests originally did just use `foreach`, and the generated `.ini` files were hardcoded in `setup_ini`. But then I was curious about whether persistent `set` and `delete` variations would also write to an `.ini` file if there was only one, so I changed it to `foreachx` in order to factor out the config chain. Turns out that because of how `write_filename` is defined:
```erlang
    WriteFile =
        case IniFiles of
            [_ | _] -> lists:last(IniFiles);
            _ -> undefined
        end,
```
that it does indeed write to a single `.ini` file configuration (further evidence that `default.ini` is not a great name).

So maybe I should just change it back now that my experiment has concluded there's no real edge case for just a single `.ini` config?

Comment (jaydoane) [reply src/config/test/config_tests.erl]: > My main beef with `foreachx` is that it becomes another eunit quirk one has to learn to be able add add or fix failing tests.

I find it annoying that it requires `Pairs::[{X::any(), ((X::any(), R::any()) -> Tests)}]` instead of (say) an `Instantiator` that returns `Pairs`. But, yeah, I agree we can keep its use to a minimum. I've changed those tests back to `foreach`.

File: src/config/src/config.erl [hunk 5]
```diff
@@ -324,26 +337,15 @@ handle_call({delete, Sec, Key, Persist, Reason}, _From, Config) ->
...
-    % from disk
-    dict:fold(
-        fun({Sec, Key} = K, V, _) ->
+    IniMap = ini_map(Config#config.ini_files),
+    % Update ets with list values, and delete those with ?DELETE markers.
+    maps:foreach(
+        fun({Sec, Key} = K, V) ->
             VExisting = get(Sec, Key, V),
-            ets:insert(?MODULE, {K, V}),
+            case V =:= ?DELETE of
```
Comment (rnewson) [inline src/config/src/config.erl:345]: `=:=` looks odd here as the only difference between it and the more obvious `==` applies to comparing floats and integers. `?DELETE` is an atom, so there is no chance of that.

File: src/config/src/config.erl [hunk 6]
```diff
@@ -400,10 +400,21 @@ is_sensitive(Section, Key) ->
         _ -> false
     end.
 
+ini_map(IniFiles) ->
```
Comment (rnewson) [inline src/config/src/config.erl:403]: nice.

**Review by rnewson — COMMENTED** [review]
I like the new approach and the switch to map from list is clearly helpful. 

bit concerned about some unrelated changes (the is_sensitive commit particularly) but they all seem fine. suggest updating the PR text/title to clarify this does more than currently stated, to help us in the future.

**Comment by jaydoane** [comment]
> bit concerned about some unrelated changes (the is_sensitive commit particularly) but they all seem fine. suggest updating the PR text/title to clarify this does more than currently stated, to help us in the future.

There are actually many "unrelated" changes to clean up the code, and generally make it easier to understand. The handling of sensitive values in particular took up a lot of room, and IMO obscured the core logic. I did make sure that all those changes are covered by tests.

File: src/docs/src/config/intro.rst [hunk 7]
```diff
@@ -126,9 +126,10 @@ by at least one space on each side. This might be useful in the ``[jwt_keys]``
 section, where base64 encoded keys may contain some ``=`` characters.
 
 .. note::
-    In case when you'd like to remove some parameter from the `default.ini`
-    without modifying that file, you may override in `local.ini`, but without
-    any value::
+    In case you'd like to remove some parameter from an `.ini` file
+    (e.g. `default.ini`) without modifying that file, you may override
+    in a subsequently loaded `.ini` file (e.g. `local.ini`), but
```
Comment (nickva) [inline src/docs/src/config/intro.rst:131]: Wonder if `may override it in a...` would sound better?

File: src/config/src/config.erl [hunk 8]
```diff
@@ -315,6 +315,19 @@ handle_call({delete, Sec, Key, Persist, Reason}, _From, Config) ->
...
+    Persist orelse
+        begin
+            IniMap = ini_map(Config#config.ini_files),
+            case maps:find({Sec, Key}, IniMap) of
+                % ?DELETE markers not inserted
+                {ok, Val} when is_list(Val) ->
+                    true = ets:insert(?MODULE, {{Sec, Key}, Val});
+                _ ->
+                    ok
+            end
```
Comment (nickva) [inline src/config/src/config.erl:329]: I wonder how this behavior would work?

If delete marker is not persisted, it seems it should only affect whether it gets written to the last file in the chain or not, otherwise the "live" value should be the same if persisted or not persisted case. With the new behavior, it seems, if `Persisted = false` we'd actually load the last non-deleted value from the .ini chain.

Comment (jaydoane) [reply src/config/src/config.erl:329]: With the current behavior, since we don't reload anything, the value is `undefined` following a delete (until reload or restart ofc).

With the new behavior, in the `Persist == true` case, it's still `undefined` due to the delete marker, so the only change is for `Persist == false` where, as you point out, it reloads the last non deleted .ini value. I demonstrated that behavior in this test:
```erlang
non_persistent_set_delete_reload_restart() ->
    Persist = false,
    ?assertEqual("local.ini", config:get("section", "file")),
    ?assertEqual(ok, config:set("section", "file", "memory", Persist)),
    ?assertEqual("memory", config:get("section", "file")),
    ?assertEqual(ok, config:delete("section", "file", Persist)),
    ?assertEqual("local.ini", config:get("section", "file")),
```
It may seem odd, but at least it's consistent now, since neither case changes after reload or restart.

File: src/docs/src/config/intro.rst (hunk 7 above)
Comment (jaydoane) [reply src/docs/src/config/intro.rst:131]: Oops, good catch. Will push a fix.

File: src/config/src/config.erl (hunk 8 above)
Comment (nickva) [reply src/config/src/config.erl:329]: The goal of the `Persist = false` I think would be do everything the same as `Persist = true` except write to file. That seems like the simplest description of its behavior. The immediate effects on the "live" config value should be exactly the same, so if a user does:

`config:set("section", "key", "value", true)` or `config:set("section", "key", "value", false)`, the live value would be `"value"`.  If they do `config:delete("section", "key", true)` or `config:delete("section", "key", false)` the live value would be the code default (config returns `undefined`).
 
After a restart or reload, the live value may change (it may be whatever was in the files) but that's something the user explicitly requested by using he `Persist = false` modifier.

Only in the `Persist = true` (the default) case, we'd want to have idempotence between any number of restarts or reloads after setting or deleting a value.

Comment (jaydoane) [reply src/config/src/config.erl:329]: That also makes sense, and has the added benefit of not requiring any behavior change to delete IIUC.

File: src/docs/src/config/intro.rst [hunk 9]
```diff
@@ -126,9 +126,10 @@ by at least one space on each side. This might be useful in the ``[jwt_keys]``
 section, where base64 encoded keys may contain some ``=`` characters.
 
 .. note::
-    In case when you'd like to remove some parameter from the `default.ini`
-    without modifying that file, you may override in `local.ini`, but without
-    any value::
+    In case you'd like to remove some parameter from an `.ini` file
+    (e.g. `default.ini`) without modifying that file, you may override
+    it in a subsequently loaded `.ini` file (e.g. `local.ini`), but
```
Comment (pgj) [inline src/docs/src/config/intro.rst:131]: Is the concept of being "subsequently loaded" clarified somewhere in the documentation?  That is, is the order of processing for `.ini` files explicitly discussed somewhere above?

Comment (jaydoane) [reply src/docs/src/config/intro.rst:131]: Yes, [here](https://docs.couchdb.org/en/stable/config/intro.html#configuration-files) is where the docs discuss loading configuration files in order. I'm very open to improved wording though...

Comment (pgj) [reply src/docs/src/config/intro.rst:131]: Ah, that is perfect, thank you.  I do not think I have problems with the wording.  I just wondered if this has been made explicit up to this point.

**Comment by jaydoane** [comment]
> nice improvements here. I tried it out in remsh also and found everything working the way we discussed.

Thanks for the review, Bob.

Unfortunately, while this PR does indeed "fix" `reload` to honor delete markers, it turns out that at least one big downstream user actually _relies_ on this bug to operate their DBaaS. I've added some details in the Additional Context of this issue: https://github.com/apache/couchdb/issues/4892

**Comment by rnewson** [comment]
hm, that's unfortunate, but not a good enough reason not fix a bug imo.

File: src/config/src/config.erl (hunk 3 above)
Comment (jaydoane) [reply src/config/src/config.erl]: I'm resolving this conversation since we decided to take the approach of removing support for delete markers altogether.

File: src/config/src/config.erl (hunk 8 above)
Comment (jaydoane) [reply src/config/src/config.erl]: > or after a delete call
But only after a persistent delete, right?

Comment (nickva) [reply src/config/src/config.erl]: I think in all cases? The only effect of `Persist=false` would be that the value is not written to file. So ets:delete from then call "load" and whatever "load" does is what we end up with?

Comment (jaydoane) [reply src/config/src/config.erl]: It's definitely cleaner code to do it in both cases, but I thought you had previously argued against reloading .ini files in the case of `Persist=false`?

Comment (jaydoane) [reply src/config/src/config.erl]: https://github.com/apache/couchdb/pull/4813/commits/d083263c17ee8002150ccfd1b0de923a0626d486 reloads for both kinds of `config:delete`, and IMO the tests seem more consistent.

File: src/config/src/config.erl [hunk 10]
```diff
@@ -340,54 +331,30 @@ handle_call({delete, Sec, Key, Persist, Reason}, _From, Config) ->
...
-        DiskKVs
+        IniMap
     ),
-    % And remove anything in ets that wasn't
-    % on disk.
+    % And remove anything in ets that wasn't on disk.
     ets:foldl(
         fun({{Sec, Key} = K, _}, _) ->
-            case dict:is_key(K, DiskKVs) of
+            case maps:is_key(K, IniMap) of
```
[the remark this answers is not in the document]
Comment (jaydoane) [reply src/config/src/config.erl]: excellent suggestion!
implemented in https://github.com/apache/couchdb/pull/4813/commits/0f727cd1ed2de5e22fbc13d5a120a2122da74d80

**Review by nickva — APPROVED** [review]
+1

Nice code simplification too!
